### DISCUSSION TITLE
Move source code to another JVM language

### DISCUSSION THREAD

**Discussion by mcpiroman** [oracle/graal#4768] [Ideas]
AFAIK the main goal of deciding to write GraalVM, the next execution environment of Java (and other languages), in the Java itself rather than c++, was to improve the safety and expressiveness of the source code, which allows for faster iterations etc. But, at least at the latter point, I do not consider Java to be significantly better than c++.

So how about considering to move a step further in that direction and choosing some newer JVM language, like Scala or Kotlin? Especially the latter, as it is was built mainly just to be the better java, while maintaining full compatibility with it. And it has quite a good tool for conversion from java.

I know there are some intrinsics for when graalvm compiles itself, but I assume they are all at the bytecode level, so change to the source should not matter, right?

**Comment by linghengqian** [comment]
- A key problem is that everything was not originally designed for other JVM languages. Now you want to use Scala in GraalVM Native Image is much more difficult than Java. Static compiler for Groovy in GraalVM's incubation phase? that is something that doesn't exist. As for Kotlin, at least before the K2 compiler came out, most people's biggest impression of Kotlin is its slow compilation speed.

- Using a JVM language other than Java would mean decoupling from the OpenJDK working group, which also conflicts with the current roadmap.
